pkg/controller/atlasproject/custom_roles_test.go

```diff
@@ -1,210 +1,194 @@
 package atlasproject
 
 import (
+	"context"
+	"net/http"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
-	"go.uber.org/zap"
+	"github.com/stretchr/testify/mock"
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+	"go.mongodb.org/atlas-sdk/v20231115008/mockadmin"
+	"go.uber.org/zap/zaptest"
 
-	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
 	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
-	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/status"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
 )
 
-func TestCalculateChanges(t *testing.T) {
-	desired := []akov2.CustomRole{
+func TestEnsureCustomRoles(t *testing.T) {
+	testRole := []akov2.CustomRole{
 		{
-			Name: "cr-1",
-		},
-		{
-			Name: "cr-3",
+			Name: "test-role",
 			InheritedRoles: []akov2.Role{
+				{Name: "role1", Database: "db1"},
+				{Name: "role2", Database: "db2"},
+			},
+			Actions: []akov2.Action{
+				{
+					Name: "action1",
+					Resources: []akov2.Resource{
+						{Cluster: pointer.MakePtr(true)},
+						{Database: pointer.MakePtr("db1")},
+					},
+				},
 				{
-					Name:     "admin",
-					Database: "test",
+					Name: "action2",
+					Resources: []akov2.Resource{
+						{
+							Database:   pointer.MakePtr("db2"),
+							Collection: pointer.MakePtr("test-collection"),
+						},
+					},
 				},
 			},
 		},
-		{
-			Name: "cr-4",
-		},
 	}
-	current := []akov2.CustomRole{
+
+	for _, tc := range []struct {
+		name string
+
+		roles []akov2.CustomRole
+
+		roleAPI *mockadmin.CustomDatabaseRolesApi
+
+		isOK bool
+	}{
 		{
-			Name: "cr-1",
+			name: "No Roles in AKO or Atlas (no op)",
+			roleAPI: func() *mockadmin.CustomDatabaseRolesApi {
+				roleAPI := mockadmin.NewCustomDatabaseRolesApi(t)
+				roleAPI.EXPECT().ListCustomDatabaseRoles(context.Background(), "").
+					Return(admin.ListCustomDatabaseRolesApiRequest{ApiService: roleAPI})
+				roleAPI.EXPECT().ListCustomDatabaseRolesExecute(mock.Anything).
+					Return(
+						[]admin.UserCustomDBRole{},
+						&http.Response{},
+						nil,
+					)
```

Comment:
These mocks are on the SDK library, which is fine here. But if we had added contract testing, which the translation layer enables, these would be a bit redundant and we could mock at the translation layer interface itself `List` instead of `ListCustomDatabaseRoles`.

BTW, I see we added the mockery mock support for the custom role service, are we using it at all here?
If not I would remove until we do (YAGNI)